### DISCUSSION TITLE
Don't know whats causing the kubernetes container to exit with code 0 randomly

### DISCUSSION THREAD

**Discussion by cyberjar09** [fastapi/fastapi#11269] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging
from typing import Annotated, Any
from uuid import UUID

import httpx
from dagster._core.instance import DagsterInstance
from fastapi import FastAPI, Request, Response
from pydantic import BaseModel
from typing_extensions import TypedDict

dagster_agent_app = FastAPI()

class DagsterInstanceSingleton:
    _self = None

    def __new__(cls):
        if cls._self is None:
            cls._self = super().__new__(cls)
        return cls._self

    def __init__(self):

        self.instance = DagsterInstance.from_config('.')


class RemoteDagsterRunResponse(BaseModel):
    dagster_run_requested: UUID
    executed_run: Any


@dagster_agent_app.post("/etl/launch/{dagster_run_id}", response_model=RemoteDagsterRunResponse)
async def etl(dagster_run_id: Annotated[UUID, "Dagster run id"]) -> RemoteDagsterRunResponse:
    executed_run = DagsterInstanceSingleton().instance.launch_run(str(dagster_run_id), None)
    return RemoteDagsterRunResponse(dagster_run_requested=dagster_run_id, executed_run=executed_run)

class HealthResponse(TypedDict):
    message: str

@dagster_agent_app.get("/health", response_model=HealthResponse)
async def health() -> HealthResponse:
    return {
        "message": "Healthy"
    }


if __name__ == "__main__":
    import uvicorn
    import os

    webserver_port: int = int(os.environ.get("WEBSERVER_PORT", "3000"))
    log_level: str = os.environ.get("WEBSERVER_LOG_LEVEL", "info").lower()

    uvicorn.run("main:dagster_agent_app", host="0.0.0.0", port=webserver_port, log_level=log_level)
```

### Description

- im trying to isolate the problem, but not sure if its coming from fastapi/uvicorn, could really use some help since im new to fastapi and python ecosystem in general so not sure if im messing something up in my code/image building
---
- I have a simple api service which receives `POST /something`, executes some sdk code then returns a response based on the SDK return
- the program runs in kubernetes with an istio sidecar
- I am seeing an intermittent situation where the fastapi container is restarting, causing dropped requests (the pod is still available with the istio sidecar forwarding requests)
- im not able to deterministically reproduce the problem yet

- when I set the uvicorn logs to trace, this is what I see
```
TRACE:    127.0.0.6:50613 - HTTP connection made
TRACE:    127.0.0.6:58511 - HTTP connection made
TRACE:    127.0.0.6:50613 - ASGI [1246] Started scope={'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('172.28.248.197', 3000), 'client': ('127.0.0.6', 50613), 'scheme': 'http', 'method': 'GET', 'root_path': '', 'path': '/health', 'raw_path': b'/health', 'query_string': b'', 'headers': '<...>', 'state': {}}
TRACE:    127.0.0.6:50613 - ASGI [1246] Send {'type': 'http.response.start', 'status': 200, 'headers': '<...>'}
TRACE:    127.0.0.6:50613 - ASGI [1246] Send {'type': 'http.response.body', 'body': '<21 bytes>'}
TRACE:    127.0.0.6:50613 - ASGI [1246] Completed
TRACE:    127.0.0.6:58511 - ASGI [1247] Started scope={'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'}, 'http_version': '1.1', 'server': ('172.28.248.197', 3000), 'client': ('127.0.0.6', 58511), 'scheme': 'http', 'method': 'GET', 'root_path': '', 'path': '/health', 'raw_path': b'/health', 'query_string': b'', 'headers': '<...>', 'state': {}}
TRACE:    127.0.0.6:58511 - ASGI [1247] Send {'type': 'http.response.start', 'status': 200, 'headers': '<...>'}
TRACE:    127.0.0.6:58511 - ASGI [1247] Send {'type': 'http.response.body', 'body': '<21 bytes>'}
TRACE:    127.0.0.6:58511 - ASGI [1247] Completed
TRACE:    127.0.0.6:50613 - HTTP connection lost
TRACE:    127.0.0.6:58511 - HTTP connection lost
INFO:     Shutting down
INFO:     Waiting for application shutdown.
TRACE:    ASGI [1] Receive {'type': 'lifespan.shutdown'}
TRACE:    ASGI [1] Send {'type': 'lifespan.shutdown.complete'}
TRACE:    ASGI [1] Completed
INFO:     Application shutdown complete.
INFO:     Finished server process [1]
```

-  I cant figure out why the container suddenly shuts down, it has a 0 exit code too
- the kubernetes events have nothing out of the ordinary, liveness probe fails because the container received a shutdown signal, then the container restarts
```
LAST SEEN   TYPE      REASON                 OBJECT                                                       MESSAGE
19m         Normal    Created                pod/dagster-agent-6b96964f6-5s4fc                            Created container dagster-agent
19m         Normal    Started                pod/dagster-agent-6b96964f6-5s4fc                            Started container dagster-agent
19m         Warning   Unhealthy              pod/dagster-agent-6b96964f6-5s4fc                            Liveness probe failed: Get "http://172.28.243.54:15020/app-health/dagster-agent/livez": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
19m         Warning   Unhealthy              pod/dagster-agent-6b96964f6-5s4fc                            Readiness probe failed: Get "http://172.28.243.54:15020/app-health/dagster-agent/readyz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
19m         Normal    Killing                pod/dagster-agent-6b96964f6-5s4fc                            Container dagster-agent failed liveness probe, will be restarted
19m         Normal    Pulled                 pod/dagster-agent-6b96964f6-5s4fc                            Container image "364869384120.dkr.ecr.ap-southeast-1.amazonaws.com/staging/dagster-agent:main-9c468eac" already present on machine
```

### relevant packages im using
fastapi = "0.104.1"
uvicorn = "0.24.0"
httpx = "0.27.0"

- the basic gist of my code is pasted in the other textbox

to run my program, I build with docker
```
FROM python:3.11.6-slim-bookworm

# Keeps Python from generating .pyc files in the container
ENV PYTHONDONTWRITEBYTECODE=1

# Turns off buffering for easier container logging
ENV PYTHONUNBUFFERED=1

COPY Pipfile Pipfile.lock ./
RUN python -m pip install --upgrade pip --no-compile --no-cache-dir
RUN pip install pipenv --no-compile --no-cache-dir && \
    pipenv install --system --deploy --ignore-pipfile && \
    pipenv --clear

WORKDIR /app
COPY ./src /app/src
# Exclude files matching the pattern `test_*.py`
RUN find src -type f -name 'test_*.py' -exec rm -f {} +

# Creates a non-root user and adds permission to access the /app folder
RUN adduser -u 5678 --disabled-password --gecos "" appuser && chown -R appuser /app
USER appuser

# During debugging, this entry point will be overridden. For more information, please refer to https://aka.ms/vscode-docker-python-debug
CMD ["python3", "src/main.py"]
```

### Operating System

Linux

### FastAPI Version

0.110.0

### Pydantic Version

2.6.3

### Python Version

3.11.6

**Comment by jgould22** [comment, marked as answer]
I am not super familiar with the dagster sdk so my first recommendation is for you to set your `etl` function to a `def` instead of an `async def`.

If there is blocking code (ie a lot of code without an await) in the dagster lib it will block the event loop and cause your application to fail to respond to health checks which could be causing k8s to shut down the container thinking it is unhealthy.

**Reply by cyberjar09** [reply]
nail on the head @jgould22 

- turns out it was my fault
- I used `async` which causes FastAPI to become single threaded since it expects `await`s in the body so it can offload the processing while continuing to serve requests
- but because the body of the controller was performing blocking I/O (network calls to k8s api server), a second request (/heath) that came in will fail causing liveness probe to fail causing the container to restart
- upon examining the dagster instance code, there isnt support for `await`, hence I removed the async on the controller
- removing `async` causes uvicron to fallback to the default of 40 available threads
- I have been stress testing the setup now and it seems to be the fix

thanks again for the support, tremendously appreciated and sorry for the wasted time :bow:
